src/cprover/cutpoint_graph.h

```diff
@@ -0,0 +1,169 @@
+//
+// Created by Yakir Vizel on 5/1/24.
+//
+
+#ifndef CBMC_CUTPOINT_GRAPH_H
+#define CBMC_CUTPOINT_GRAPH_H
+
+#include "goto_model.h"
+#include <vector>
+
+/**
+ * A cut-point graph over goto programs.
+ * The cut-points are:
+ * 1. Entry node
+ * 2. Exit node
+ * 3. Every node that has a back-edge - namely, a loop head.
+ */
+
+class cutpoint_graph;
+class cutpoint;
+
+class cutpoint_edge
+{
+  friend class cutpoint_graph;
+
+  cutpoint & m_src;
+  cutpoint & m_dst;
+
+  typedef std::vector<const goto_programt::instructiont *> InstVec;
+  InstVec m_insts;
+
+  cutpoint &source() { return m_src; }
+  cutpoint &target() { return m_dst; }
+public:
+  cutpoint_edge(cutpoint & s, cutpoint & d) : m_src(s), m_dst(d) {}
+
+  const cutpoint_graph& graph() const;
+
+  const cutpoint &source() const { return m_src; }
+  const cutpoint &target() const { return m_dst; }
+
+  void push_back(const goto_programt::instructiont * inst)
+  {
+    m_insts.push_back(inst);
+  }
+
+  typedef InstVec::iterator iterator;
+  typedef InstVec::const_iterator const_iterator;
+
+  iterator begin() { return m_insts.begin(); }
+  iterator end() { return m_insts.end(); }
+  const_iterator begin() const { return m_insts.begin(); }
+  const_iterator end() const { return m_insts.end(); }
+};
+
+typedef std::shared_ptr<cutpoint_edge> cutpoint_edge_ptr;
+
+class cutpoint
+{
+  friend class cutpoint_graph;
+
+  const cutpoint_graph &m_graph;
+  unsigned m_id;
```

Comment:
`unsigned` will hardly ever be the right type. Should it perhaps be `std::size_t`?